lib/livebook_web/live/hooks/app_auth_hook.ex

```diff
@@ -0,0 +1,88 @@
+defmodule LivebookWeb.AppAuthHook do
+  import Phoenix.Component
+  import Phoenix.LiveView
+
+  # For apps with password, we want to store the hashed password
+  # (let's call it token) in the session, as we do for the main auth.
+  # However, the session uses cookies, which have a ~4kb size limit.
+  # We could use multiple cookies, however there are also limits on
+  # the number of cookies, and we don't want the browser to clear
+  # them all at some point. Additionally, accumulating cookies for
+  # all apps would imply larger payloads on every regular request.
+  #
+  # Since we don't have any persistence on the server side, the other
+  # option is to use the browser local storage and manage it through
+  # JavaScript. Therefore, the whole auth is built using LiveView.
+  # The flows are:
+  #
+  #   * unauthenticated - the auth LiveView shows a regular form and
+  #     validates the user-provided password. Once the password is
+  #     correct, it pushes an event to the client to store the token.
+  #     Once the token is stored it redirects to the app page.
+  #
+  #   * authenticated - on dead render the app LiveView renders just
+  #     a loading screen. On the client side, provided it's the app
+  #     page, we read the token from local storage (if stored) and
+  #     send it in mount connect params via the socket. Then on the
+  #     server we use that token to authenticate.
+  #
+  # This module defines a hook that sets the `:app_authenticated?`
+  # assign to reflect the current authentication state. For public
+  # apps (or in case the user has full access) it is set to `true`
+  # on both dead and live render.
+
+  def on_mount(:default, %{"slug" => slug}, session, socket) do
+    case Livebook.Apps.fetch_settings_by_slug(slug) do
+      {:ok, %{access_type: :public}} ->
+        {:cont, assign(socket, app_authenticated?: true)}
```

Comment:
Let's also store the app settings in the assign, since we use it in both AppAuth and AppLive anyway. Another idea is to store only the app_settings and check if it is nil or not. :)

Reply:
Good call!